### ISSUE TITLE
Reclaiming disk space from job table

### ISSUE THREAD

**Issue by Sinjo** [chanks/que#90]
As things are, Que uses a single table (`que_jobs`) to store jobs. This makes it difficult to get disk space back from a busy queue.

The reason for this is that Postgres [doesn't release disk space back to the OS](http://www.postgresql.org/docs/9.2/static/routine-vacuuming.html#VACUUM-FOR-SPACE-RECOVERY) during a normal `VACUUM`<sup>1</sup>.

One approach I've seen used elsewhere<sup>2</sup> is to have multiple job tables, rotate between them, and truncate old ones once all the jobs in them have been processed. Running `TRUNCATE` on a table releases space back to the OS.

Under normal conditions I wouldn't expect this to be too much of an issue. Postgres will clear and reuse old pages in the table. The problem comes if you have a large spike in unprocessed jobs (e.g. your workers fall over and it takes time to recover them). Getting that disk space back on a high-traffic system becomes difficult.

I was wondering if you had any thoughts on this. Table rotation is a bit of a pain<sup>3</sup>, but I don't know of any other approach to solving this issue.

[1] There's one small exception to this: if there are free pages at the end of a table, and the vacuum process can easily acquire an `ACCESS EXCLUSIVE` lock, those will be released to the OS.

[2] Specifically in [PGQ](https://wiki.postgresql.org/wiki/SkyTools#PgQ)

[3] There's coordination involved: you want to be sure that new jobs are only going to the new job table before you truncate the old one, and you want to make sure your workers will drain the old queue before switching over to the new one.

**Comment by chanks** [comment]
I'm curious to hear whether anyone else has had this as an issue. Any kind of table rotation scheme that was built into Que would become complicated, to put it mildly, and would only really benefit users who:

1) Had a large, temporary spike in the size of their jobs table.
2) Have a sufficiently small amount of disk space that the size of the bloated jobs table is a serious issue.
3) Have such stringent uptime/throughput requirements that a VACUUM FULL of the jobs table isn't a possibility.

I don't think this is terribly common. Is this a theoretical issue or one you've actually run into?

**Comment by Sinjo** [comment]
I'll admit this one's theoretical. Right now I'm looking through a few Postgres-backed queues to use on a project. Que and PGQ are top of the list (already ruled out DelayedJob and QueueClassic).

3 is very true for us. We want to enqueue things on the critical path (i.e. as part of an API request), and downtime isn't an option there.

For 1 and 2, I guess it depends how big the spike is. Right now I'm wondering whether the caveat on `VACUUM` freeing pages at the end of the table is enough to shrink the table to a manageable size. If new jobs get written into older, freed up pages, the next `VACUUM` along should be able to release space at the end of the table back to the OS, provided the vacuum daemon can acquire an exclusive lock. I'm not sure how likely that is to work in practice (any writes/reads on the table will prevent that exclusive lock being taken).

I agree with you on keeping complexity down. That's actually one of the things I really like about Que over PGQ: I feel like the codebase is a lot more approachable. I'm just also keen to make sure it behaves well over time, including the ability to recover from failures.

**Comment by chanks** [comment]
The thing about VACUUM freeing pages at the end of the table being enough feels right to me, but this isn't a problem I've run into, so I don't know. I believe the slowest thing about VACUUM FULL will be rewriting the jobs into a new table, which should be very fast if you don't have many jobs. So, the smaller your current dataset (and the more space you have to reclaim), the faster it should be. I may be wrong, but it should be easy to benchmark that if you're curious.

**Comment by Sinjo** [comment]
Yeah, that sounds right. I'll report back here with any more benchmarks/experiments I've run. Totally behind keeping complexity out unless it proves necessary.

**Comment by Sinjo** [comment]
I did some extremely high-level testing locally, and it looks like that caveat on regular `VACUUM` works as expected. My setup was:

* A Rails console enqueuing jobs in a tight loop - `loop { Workers::TestQueJob.enqueue }`
* A query being run every 2 seconds to measure the on-disk size of the `que_jobs` table - `SELECT pg_relation_filepath(oid), relpages FROM pg_class WHERE relname = 'que_jobs';`
* A `VACUUM` query being run every 2 seconds - `VACUUM que_jobs;`

First up, I deleted everything except the last job which was written. As expected the on-disk size didn't drop<sup>1</sup>. After about 10 seconds, I deleted that job. At that point, the next `VACUUM` reduced the on-disk size a lot.

I'm not going to claim this represents a production system - this is a throwaway local test, and what that looks like will vary wildly based on the app. This test makes me fairly comfortable that most apps will be fine though. You'd have to have high contention on the jobs table at all times for this not to work.

Going with the current approach does make it more critical to think about your vacuuming setup, but I think that's a fair tradeoff to make for reducing the complexity of Que.

I'll report back here if I do any more rigorous benchmarking, or if this does turn out to be a real issue for us.

[1] `VACUUM` operations running at this point will have freed up early pages in the table for reuse.
